cpp/src/copying/gather.cu

```diff
@@ -51,224 +79,102 @@ struct bounds_checker {
-void gather_bitmask(gdf_valid_type const* source_mask,
-                    gdf_size_type num_source_rows,
-                    gdf_valid_type* destination_mask,
-                    gdf_size_type num_destination_rows,
-                    gdf_index_type const gather_map[],
-                    bool check_bounds = false, cudaStream_t stream = 0) {
-  CUDF_EXPECTS(destination_mask != nullptr, "Missing valid buffer allocation");
-
-  constexpr gdf_size_type BLOCK_SIZE{256};
-  const gdf_size_type gather_grid_size =
-      (num_destination_rows + BLOCK_SIZE - 1) / BLOCK_SIZE;
-
-  gdf_valid_type* output_bitmask{destination_mask};
-
-  // Allocate a temporary results buffer if gathering in-place
-  bool const in_place{source_mask == destination_mask};
-  rmm::device_vector<gdf_valid_type> temp_bitmask;
-  if (in_place) {
-    temp_bitmask.resize(gdf_valid_allocation_size(num_destination_rows));
-    output_bitmask = temp_bitmask.data().get();
-  }
-
-  if (check_bounds) {
-    gather_bitmask_if_kernel<<<gather_grid_size, BLOCK_SIZE, 0, stream>>>(
-        source_mask, num_source_rows, output_bitmask, num_destination_rows,
-        gather_map, gather_map, bounds_checker{0, num_source_rows});
-  } else {
-    gather_bitmask_kernel<<<gather_grid_size, BLOCK_SIZE, 0, stream>>>(
-        source_mask, num_source_rows, output_bitmask, num_destination_rows,
-        gather_map);
-  }
-
-  CHECK_STREAM(stream);
+  return f;
+}
 
-  if (in_place) {
-    thrust::copy(rmm::exec_policy(stream)->on(stream), temp_bitmask.begin(),
-                 temp_bitmask.end(), destination_mask);
+template <bool check_bounds>
+__global__ void gather_bitmask_kernel(const bit_mask_t* const* source_valid,
+                                      gdf_size_type num_source_rows,
+                                      const gdf_index_type* gather_map,
+                                      bit_mask_t** destination_valid,
+                                      gdf_size_type num_destination_rows,
+                                      gdf_size_type* d_count,
+                                      gdf_size_type num_columns) {
+  for (gdf_index_type i = 0; i < num_columns; i++) {
+    const bit_mask_t* source_valid_col = source_valid[i];
+    bit_mask_t* destination_valid_col = destination_valid[i];
+
+    const bool src_has_nulls = source_valid_col != nullptr;
+    const bool dest_has_nulls = destination_valid_col != nullptr;
+
+    if (dest_has_nulls) {
+      gdf_index_type destination_row_base = blockIdx.x * blockDim.x;
+
+      gdf_size_type valid_count_accumulate = 0;
+
+      while (destination_row_base < num_destination_rows) {
+        gdf_index_type destination_row = destination_row_base + threadIdx.x;
+
+        const bool thread_active = destination_row < num_destination_rows;
+        gdf_index_type source_row =
+            thread_active ? gather_map[destination_row] : 0;
+
+        const uint32_t active_threads =
```

Comment:
You need to compute this before the `while()`, otherwise the `0xffffffff` mask is incorrect because some threads do not enter the `while()`.

Reply:
The variable used in `while` is `destination_row_base` instead of `destination_row` so it works if the block_size is a multiple of `warp_size`. But it does make people uncomfortable as it's not the usual practice. Will change that.

Reply:
Ah, yes, okay. This makes sense.